osquery/tables/utility/file.cpp

```diff
@@ -25,8 +32,144 @@ namespace osquery {
 
 namespace tables {
 
-#if !defined(WIN32)
+namespace {
+#ifdef WIN32
+
+struct LnkData {
+  std::string target_path;
+  std::string target_type;
+  std::string target_location;
+  std::string start_in;
+  std::string run;
+  std::string comment;
+};
+
+std::string showCmdToString(int show_cmd) {
+  switch (show_cmd) {
+  case SW_SHOWNORMAL: {
+    return "Normal window";
+  }
+  case SW_SHOWMAXIMIZED: {
+    return "Maximized";
+  }
+  case SW_SHOWMINIMIZED: {
+    return "Minimized";
+  }
+  default: {
+    return "Unknown";
+  }
+  }
+}
+
+boost::optional<LnkData> parseLnkData(const fs::path& link) {
+  IShellLink* shell_link;
+  auto hres = CoCreateInstance(CLSID_ShellLink,
+                               nullptr,
+                               CLSCTX_INPROC_SERVER,
+                               IID_IShellLink,
+                               reinterpret_cast<LPVOID*>(&shell_link));
+  if (FAILED(hres)) {
+    TLOG << "Failed to create an instance of a shell link, error: "
+         << shell_link;
+    return boost::none;
+  }
+
+  IPersistFile* file;
+  hres = shell_link->QueryInterface(IID_IPersistFile,
+                                    reinterpret_cast<LPVOID*>(&file));
+
+  if (FAILED(hres)) {
+    TLOG << "Failed to create an instance of a shell link, error: "
+         << shell_link;
+    return boost::none;
+  }
+
+  hres = file->Load(link.c_str(), STGM_READ);
+
+  if (FAILED(hres)) {
+    // Not a shell link
+    return boost::none;
+  }
+
+  /* Empty files are still able to be loaded via the ShellLink COM interface,
+     but they are not ShellLink files, so verify that the file
+     contains a header of a certain size */
+  std::string header_size;
+  auto status = readFile(link, header_size, 4);
+
+  if (!status.ok() || header_size.size() != 4) {
+    return boost::none;
+  }
+
+  std::uint32_t expected_size = 0x4C;
```

Comment:
I'm a bit confused about this being named _size. Is this an expected header size, or an expected value (which is how I read the actual code)?

Reply:
Yeah the name is unfortunate, the value represents the expected header size which we expect to read as a value in the first bytes of the file header.
I'll add this as a global constant with a clearer name.

Reply:
I'll also change the `memcmp` later to a `memcpy` to a new variable + if comparison, just to be more transparent.
I was being too clever here.